MdeModulePkg/Bus/Pci/NvmExpressDxe/NvmExpressHci.c

```diff
@@ -918,6 +1019,21 @@ NvmeControllerInit (
   DEBUG ((DEBUG_INFO, "    CQES      : 0x%x\n", Private->ControllerData->Cqes));
   DEBUG ((DEBUG_INFO, "    NN        : 0x%x\n", Private->ControllerData->Nn));
 
+  // MU_CHANGE [BEGIN] - Request Number of Queues from Controller
+  //
+  // Send Set Features Command to request the maximum number of data queues.
+  // The controller is free to allocate a different number of queues from the number requested.
+  // The number of queues allocated is returned and stored in the controller private data structure
+  // using the NumberOfDataQueuePairs field.
+  //
+  Status = NvmeSetFeaturesNumberOfQueues (Private, NVME_MAX_QUEUES - 1);

```

Comment:
Aren't you double subtracting one? You have MAX_QUEUES is 3, -1 here, which gets 2 passed into the func, but then in the func you have:

```
// Populate the Set Features Cdw10 and Cdw11 according to Nvm Express 1.3d Spec
  // Note we subtract 1 from the requested number of queues to get the 0-based value
  SetFeatures.Bits.Fid             = NVME_FEATURE_NUMBER_OF_QUEUES;
  NumberOfQueuesRequested.Bits.Ncq = Ndqpr -1;
  NumberOfQueuesRequested.Bits.Nsq = Ndqpr -1;
  ```
  
  So the number of queues requested ends up as 1? Instead of 2?

Reply:
NVME_MAX_QUEUES includes the admin queues, so -1 to get to 2 data queue pairs. 
The SetFeatures command uses 0-based values so 0 = 1 data queue pair, 1=2 data queue pairs, etc.
I just added a check for the parameter being 0 to return invalid parameter.

Reply:
Gotcha, I was missing that this didn't include the admin queue in one spot but does in the other.